addons/mrp/models/mrp_production.py

```diff
@@ -2086,6 +2136,117 @@ def action_merge(self):
     def _has_workorders(self):
         return self.workorder_ids
 
+    def _link_bom(self, bom):
+        """ Links the given BoM to the MO. Assigns BoM's lines, by-products and operations
+        to the corresponding MO's components, by-products and workorders.
+        This method assumes the production isn't linked to any BoM yet.
+        """
+        self.ensure_one()
+        moves_to_unlink = self.env['stock.move']
+        workorders_to_unlink = self.env['mrp.workorder']
+        # For draft MO, all the work will be done by compute methods.
+        # For cancelled and done MO, we don't want to do anything more than assinging the BoM.
+        if self.state == 'draft' and self.bom_id == bom:
+            # Empties `bom_id` field so when the BoM is reassigns to this field, depending computes
+            # will be triggered (doesn't happen if the field's value doesn't change).
+            self.bom_id = False
+        if self.state in ['cancel', 'done', 'draft']:
+            if self.state == 'draft':
+                # Don't straight delete the moves/workorders to avoid to cancel the MO, those will
+                # be deleted once the BoM is assigned (and thus after new moves/WO were created).
+                moves_to_unlink = self.move_raw_ids
+                workorders_to_unlink = self.workorder_ids
+            self.bom_id = bom
+            moves_to_unlink.unlink()
+            workorders_to_unlink.unlink()
+            return
+
+        def operation_key_values(record):
+            return tuple(record[key] for key in ('company_id', 'name', 'workcenter_id'))
+
+        product_attribute_ids = self.product_id.product_template_attribute_value_ids.ids
+        ratio = bom.product_qty / self.product_qty
```

Comment:
Because in this commit, I need to get the ratio only here.
In the following commit, I need to get the ratio in another place too, and thus, to avoid code duplication, I added the `_get_ratio_between_mo_and_bom_quantities` method.